presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -278,9 +286,21 @@ private Type getCommonSuperTypeForDecimal(DecimalType firstType, DecimalType sec
         return createDecimalType(targetPrecision, targetScale);
     }
 
-    private Type getCommonSuperTypeForVarchar(VarcharType firstType, VarcharType secondType)
+    private Type getCommonSuperTypeForCharacterStrings(Type firstType, Type secondType)
+    {
+        int returnTypeLength = Math.max(getCharacterStringTypeLength(firstType), getCharacterStringTypeLength(secondType));
+        if (firstType instanceof VarcharType || secondType instanceof VarcharType) {
+            return createVarcharType(returnTypeLength);
+        }
+        return createCharType(returnTypeLength);
```

Comment:
Actually, according to SQL spec: https://mariadb.com/kb/en/sql-99/character-string-operations/#assignment, it should be possible to cast both ways (`shorter -> longer`, `longer -> shorter`). There should be a warning if truncate occurs. For sake of compatibility with `VARCHAR`, I would leave it as is (e.g: `shorter -> longer`). What do you think?